File: frb_example/pure_dart/lib/src/rust/api/exception.dart
// ignore_for_file: invalid_use_of_internal_member, unused_import

import '../frb_generated.dart';
import 'package:flutter_rust_bridge/flutter_rust_bridge_for_generated.dart';
import 'package:freezed_annotation/freezed_annotation.dart' hide protected;
part 'exception.freezed.dart';

Future<int> funcReturnErrorTwinNormal({dynamic hint}) =>
    RustLib.instance.api.funcReturnErrorTwinNormal(hint: hint);

Future<int> funcTypeFalliblePanicTwinNormal({dynamic hint}) =>
    RustLib.instance.api.funcTypeFalliblePanicTwinNormal(hint: hint);

Future<int> funcTypeInfalliblePanicTwinNormal({dynamic hint}) =>
    RustLib.instance.api.funcTypeInfalliblePanicTwinNormal(hint: hint);

Future<int> customEnumErrorReturnOkTwinNormal(
        {required int arg, dynamic hint}) =>
    RustLib.instance.api
        .customEnumErrorReturnOkTwinNormal(arg: arg, hint: hint);

Future<void> customEnumErrorPanicTwinNormal({dynamic hint}) =>
    RustLib.instance.api.customEnumErrorPanicTwinNormal(hint: hint);

Future<int> customEnumErrorReturnErrorTwinNormal({dynamic hint}) =>
    RustLib.instance.api.customEnumErrorReturnErrorTwinNormal(hint: hint);

Future<void> customNestedErrorReturnErrorTwinNormal(
        {required CustomNestedErrorOuterTwinNormal arg, dynamic hint}) =>
    RustLib.instance.api
        .customNestedErrorReturnErrorTwinNormal(arg: arg, hint: hint);

Future<void> customStructErrorReturnErrorTwinNormal(
        {required CustomStructErrorTwinNormal arg, dynamic hint}) =>
    RustLib.instance.api
        .customStructErrorReturnErrorTwinNormal(arg: arg, hint: hint);

Future<int> returnErrCustomError({dynamic hint}) =>
    RustLib.instance.api.returnErrCustomError(hint: hint);

Future<int> returnOkCustomError({dynamic hint}) =>
    RustLib.instance.api.returnOkCustomError(hint: hint);

Future<int> returnErrorVariant({required int variant, dynamic hint}) =>
    RustLib.instance.api.returnErrorVariant(variant: variant, hint: hint);

Future<void> returnCustomNestedError1({dynamic hint}) =>
    RustLib.instance.api.returnCustomNestedError1(hint: hint);

Future<void> returnCustomNestedError1Variant1({dynamic hint}) =>
    RustLib.instance.api.returnCustomNestedError1Variant1(hint: hint);

Future<void> returnCustomNestedError2({dynamic hint}) =>
    RustLib.instance.api.returnCustomNestedError2(hint: hint);

Future<void> returnCustomStructError({dynamic hint}) =>
    RustLib.instance.api.returnCustomStructError(hint: hint);

void syncReturnCustomStructError({dynamic hint}) =>
    RustLib.instance.api.syncReturnCustomStructError(hint: hint);

Future<int> returnCustomStructOk({dynamic hint}) =>
    RustLib.instance.api.returnCustomStructOk(hint: hint);

Future<void> throwAnyhow({dynamic hint}) =>
    RustLib.instance.api.throwAnyhow(hint: hint);

Future<void> panicWithCustomResult({dynamic hint}) =>
    RustLib.instance.api.panicWithCustomResult(hint: hint);

Stream<String> streamSinkThrowAnyhow({dynamic hint}) =>
    RustLib.instance.api.streamSinkThrowAnyhow(hint: hint);

@freezed
sealed class CustomEnumErrorTwinNormal
    with _$CustomEnumErrorTwinNormal
    implements FrbException {
  @Implements<FrbBacktracedException>()
  const factory CustomEnumErrorTwinNormal.one({
    required String message,
    required String backtrace,
  }) = CustomEnumErrorTwinNormal_One;
  @Implements<FrbBacktracedException>()
  const factory CustomEnumErrorTwinNormal.two({
    required int message,
    required String backtrace,
  }) = CustomEnumErrorTwinNormal_Two;
}

@freezed
sealed class CustomError with _$CustomError implements FrbException {
  @Implements<FrbBacktracedException>()
  const factory CustomError.error0({
    required String e,
    required String backtrace,
  }) = CustomError_Error0;
  @Implements<FrbBacktracedException>()
  const factory CustomError.error1({
    required int e,
    required String backtrace,
  }) = CustomError_Error1;
}

@freezed
sealed class CustomNestedError1
    with _$CustomNestedError1
    implements FrbException {
  const factory CustomNestedError1.customNested1(
    String field0,
  ) = CustomNestedError1_CustomNested1;
  const factory CustomNestedError1.errorNested(
    CustomNestedError2 field0,
  ) = CustomNestedError1_ErrorNested;
}

@freezed
sealed class CustomNestedError2 with _$CustomNestedError2 {
  const factory CustomNestedError2.customNested2(
    String field0,
  ) = CustomNestedError2_CustomNested2;
  const factory CustomNestedError2.customNested2Number(
    int field0,
  ) = CustomNestedError2_CustomNested2Number;
}

@freezed
sealed class CustomNestedErrorInnerTwinNormal
    with _$CustomNestedErrorInnerTwinNormal {
  const factory CustomNestedErrorInnerTwinNormal.three(
    String field0,
  ) = CustomNestedErrorInnerTwinNormal_Three;
  const factory CustomNestedErrorInnerTwinNormal.four(
    int field0,
  ) = CustomNestedErrorInnerTwinNormal_Four;
}

@freezed
sealed class CustomNestedErrorOuterTwinNormal
    with _$CustomNestedErrorOuterTwinNormal {
  const factory CustomNestedErrorOuterTwinNormal.one(
    String field0,
  ) = CustomNestedErrorOuterTwinNormal_One;
  const factory CustomNestedErrorOuterTwinNormal.two(
    CustomNestedErrorInnerTwinNormal field0,
  ) = CustomNestedErrorOuterTwinNormal_Two;
}

class CustomStruct {
  final String message;

  const CustomStruct({
    required this.message,
  });

  static Future<CustomStruct> newCustomStruct(
          {required String message, dynamic hint}) =>
      RustLib.instance.api.customStructNew(message: message, hint: hint);

  Future<void> nonstaticReturnCustomStructError({dynamic hint}) =>
      RustLib.instance.api.customStructNonstaticReturnCustomStructError(
        that: this,
      );

  Future<int> nonstaticReturnCustomStructOk({dynamic hint}) =>
      RustLib.instance.api.customStructNonstaticReturnCustomStructOk(
        that: this,
      );

  static Future<void> staticReturnCustomStructError({dynamic hint}) =>
      RustLib.instance.api
          .customStructStaticReturnCustomStructError(hint: hint);

  static Future<int> staticReturnCustomStructOk({dynamic hint}) =>
      RustLib.instance.api.customStructStaticReturnCustomStructOk(hint: hint);

  @override
  int get hashCode => message.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CustomStruct &&
          runtimeType == other.runtimeType &&
          message == other.message;
}

class CustomStructError implements FrbException {
  final String message;

  const CustomStructError({
    required this.message,
  });

  @override
  int get hashCode => message.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CustomStructError &&
          runtimeType == other.runtimeType &&
          message == other.message;
}

class CustomStructErrorTwinNormal {
  final String a;

  const CustomStructErrorTwinNormal({
    required this.a,
  });

  @override
  int get hashCode => a.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is CustomStructErrorTwinNormal &&
          runtimeType == other.runtimeType &&
          a == other.a;
}

class SomeStruct {
  final int value;

  const SomeStruct({
    required this.value,
  });

  static Future<SomeStruct> newSomeStruct({required int value, dynamic hint}) =>
      RustLib.instance.api.someStructNew(value: value, hint: hint);

  Future<int> nonStaticReturnErrCustomError({dynamic hint}) =>
      RustLib.instance.api.someStructNonStaticReturnErrCustomError(
        that: this,
      );

  Future<int> nonStaticReturnOkCustomError({dynamic hint}) =>
      RustLib.instance.api.someStructNonStaticReturnOkCustomError(
        that: this,
      );

  static Future<int> staticReturnErrCustomError({dynamic hint}) =>
      RustLib.instance.api.someStructStaticReturnErrCustomError(hint: hint);

  static Future<int> staticReturnOkCustomError({dynamic hint}) =>
      RustLib.instance.api.someStructStaticReturnOkCustomError(hint: hint);

  @override
  int get hashCode => value.hashCode;

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is SomeStruct &&
          runtimeType == other.runtimeType &&
          value == other.value;
}

File: frb_example/pure_dart/rust/src/api/exception.rs
use anyhow::{anyhow, Result};
use backtrace::Backtrace;
use flutter_rust_bridge::StreamSink;

// ------------------------------ built-in errors ----------------------------------

pub fn func_return_error_twin_normal() -> Result<i32> {
    Err(anyhow!("deliberate error"))
}

pub fn func_type_fallible_panic_twin_normal() -> Result<i32> {
    panic!("deliberate panic")
}

pub fn func_type_infallible_panic_twin_normal() -> i32 {
    panic!("deliberate panic")
}

// ------------------------------ custom error + return ok/panic ----------------------------------

pub fn custom_enum_error_return_ok_twin_normal(arg: u32) -> Result<u32, CustomEnumErrorTwinNormal> {
    Ok(arg)
}

pub fn custom_enum_error_panic_twin_normal() -> Result<(), CustomEnumErrorTwinNormal> {
    panic!("deliberate panic");
}

// ------------------------------ custom struct error ----------------------------------

pub enum CustomEnumErrorTwinNormal {
    One {
        message: String,
        backtrace: Backtrace,
    },
    Two {
        message: u32,
        backtrace: Backtrace,
    },
}

pub fn custom_enum_error_return_error_twin_normal() -> Result<u32, CustomEnumErrorTwinNormal> {
    Err(CustomEnumErrorTwinNormal::One {
        message: "deliberate error".into(),
        backtrace: Backtrace::new(),
    })
}

// ------------------------------ custom nested errors ----------------------------------

pub enum CustomNestedErrorOuterTwinNormal {
    One(String),
    Two(CustomNestedErrorInnerTwinNormal),
}

pub enum CustomNestedErrorInnerTwinNormal {
    Three(String),
    Four(u32),
}

pub fn custom_nested_error_return_error_twin_normal(
    arg: CustomNestedErrorOuterTwinNormal,
) -> Result<(), CustomNestedErrorOuterTwinNormal> {
    Err(arg)
}

// ------------------------------ custom struct errors ----------------------------------

pub struct CustomStructErrorTwinNormal {
    pub a: String,
}

pub fn custom_struct_error_return_error_twin_normal(
    arg: CustomStructErrorTwinNormal,
) -> Result<(), CustomStructErrorTwinNormal> {
    Err(arg)
}

// ------------------------------ example-based ----------------------------------

pub enum CustomError {
    Error0 { e: String, backtrace: Backtrace },
    Error1 { e: u32, backtrace: Backtrace },
}

pub fn return_err_custom_error() -> Result<u32, CustomError> {
    Err(CustomError::Error0 {
        e: "".into(),
        backtrace: Backtrace::new(),
    })
}

pub fn return_ok_custom_error() -> Result<u32, CustomError> {
    Ok(3)
}

pub fn return_error_variant(variant: u32) -> Result<u32, CustomError> {
    match variant {
        0 => Err(CustomError::Error0 {
            e: "variant0".to_string(),
            backtrace: Backtrace::new(),
        }),
        1 => Err(CustomError::Error1 {
            e: 1,
            backtrace: Backtrace::new(),
        }),
        _ => panic!("unsupported variant"),
    }
}

pub struct SomeStruct {
    pub value: u32,
}

impl SomeStruct {
    pub fn new(value: u32) -> SomeStruct {
        SomeStruct { value }
    }

    pub fn static_return_err_custom_error() -> Result<u32, CustomError> {
        Err(CustomError::Error1 {
            e: 3,
            backtrace: Backtrace::new(),
        })
    }

    pub fn static_return_ok_custom_error() -> Result<u32, CustomError> {
        Ok(3)
    }

    pub fn non_static_return_err_custom_error(&self) -> Result<u32, CustomError> {
        Err(CustomError::Error1 {
            e: self.value,
            backtrace: Backtrace::new(),
        })
    }

    pub fn non_static_return_ok_custom_error(&self) -> Result<u32, CustomError> {
        Ok(self.value)
    }
}

pub enum CustomNestedError1 {
    CustomNested1(String),
    ErrorNested(CustomNestedError2),
}

pub enum CustomNestedError2 {
    CustomNested2(String),
    CustomNested2Number(u32),
}

pub fn return_custom_nested_error_1() -> Result<(), CustomNestedError1> {
    Err(CustomNestedError1::ErrorNested(
        CustomNestedError2::CustomNested2Number(3),
    ))
}

pub fn return_custom_nested_error_1_variant1() -> Result<(), CustomNestedError1> {
    Err(CustomNestedError1::CustomNested1("custom".to_string()))
}

pub fn return_custom_nested_error_2() -> Result<(), CustomNestedError2> {
    Err(CustomNestedError2::CustomNested2("custom".to_string()))
}
pub struct CustomStructError {
    pub message: String,
}

pub fn return_custom_struct_error() -> Result<(), CustomStructError> {
    Err(CustomStructError {
        message: "error message".to_string(),
    })
}

#[flutter_rust_bridge::frb(sync)]
pub fn sync_return_custom_struct_error() -> Result<(), CustomStructError> {
    Err(CustomStructError {
        message: "error message".to_string(),
    })
}

pub fn return_custom_struct_ok() -> Result<u32, CustomStructError> {
    Ok(3)
}

pub struct CustomStruct {
    pub message: String,
}

impl CustomStruct {
    pub fn new(message: String) -> CustomStruct {
        CustomStruct { message }
    }

    pub fn static_return_custom_struct_error() -> Result<(), CustomStructError> {
        Err(CustomStructError {
            message: "error message".to_string(),
        })
    }

    pub fn static_return_custom_struct_ok() -> Result<u32, CustomStructError> {
        Ok(3)
    }

    pub fn nonstatic_return_custom_struct_error(&self) -> Result<(), CustomStructError> {
        Err(CustomStructError {
            message: self.message.clone(),
        })
    }

    pub fn nonstatic_return_custom_struct_ok(&self) -> Result<u32, CustomStructError> {
        Ok(3)
    }
}

pub fn throw_anyhow() -> Result<(), anyhow::Error> {
    Err(anyhow!("anyhow error"))
}

pub fn panic_with_custom_result() -> Result<(), CustomError> {
    panic!("just a panic");
}

pub fn stream_sink_throw_anyhow(_sink: StreamSink<String>) -> Result<()> {
    Err(anyhow!("anyhow error"))
}
